src/tree_sortable.rs

```
@@ -0,0 +1,146 @@
+    // -- manual --
+    fn set_default_sort_func<F>(&self, sort_func: F)
+        where F: Fn(&Self, &TreeIter, &TreeIter) -> Ordering + 'static;
+    fn set_sort_func<F>(&self, sort_column_id: SortColumn, sort_func: F)
+        where F: Fn(&Self, &TreeIter, &TreeIter) -> Ordering + 'static;
+    fn get_sort_column_id(&self) -> Option<(SortColumn, SortType)>;
+    fn set_sort_column_id(&self, sort_column_id: SortColumn, order: SortType);
+    fn set_unsorted(&self);
+    // -- auto --
+    fn has_default_sort_func(&self) -> bool;
+    fn sort_column_changed(&self);
+}
+
+unsafe extern "C" fn trampoline<T>(this: *mut GtkTreeSortable, iter: *mut GtkTreeIter,
+                                   iter2: *mut GtkTreeIter,
+                                   f: gpointer) -> i32
+where T: IsA<TreeSortable> {
+    callback_guard!();
+    let f : &Box<Fn(&T, &TreeIter, &TreeIter) -> Ordering> = transmute(f);
+    f(&TreeSortable::from_glib_none(this).downcast_unchecked(), &from_glib_borrow(iter),
+      &from_glib_borrow(iter2)).to_glib()
+}
+
+unsafe extern "C" fn destroy_closure(ptr: gpointer) {
+    callback_guard!();
+    Box::<Box<Fn(&TreeSortable, &TreeIter, &TreeIter) -> Ordering + 'static>>::from_raw(ptr as *mut _);
+}
+
+fn into_raw<F, T>(func: F) -> gpointer
+    where F: Fn(&T, &TreeIter, &TreeIter) -> Ordering + 'static {
+    let func: Box<Box<Fn(&T, &TreeIter, &TreeIter) -> Ordering + 'static>> =
+        Box::new(Box::new(func));
+    Box::into_raw(func) as gpointer
+}
+
+impl<O: IsA<TreeSortable>> TreeSortableExt for O {
+    #[inline]
+    fn get_sort_column_id(&self) -> Option<(SortColumn, SortType)> {
+        unsafe {
+            let mut sort_column_id = mem::uninitialized();
+            let mut order = mem::uninitialized();
+            let ret = from_glib(ffi::gtk_tree_sortable_get_sort_column_id(self.to_glib_none().0, &mut sort_column_id, &mut order));
+            if ret || sort_column_id != ffi::GTK_TREE_SORTABLE_UNSORTED_SORT_COLUMN_ID {
```

If (unlikely) a new variant (e.g. `-3`) appears this will fail silently instead of tripping the assertion.